File: backend/models/Usuario.js
// models/Usuario.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const Reserva = require('./Reserva');
const PlanDieta = require('./PlanDieta');
const PlanEntrenamiento = require('./PlanEntrenamiento');
const Salud = require('./Salud');
const esquemaUsuario = new mongoose.Schema({
  nombre: { type: String, required: true, trim: true },
  correo: { type: String, required: true, unique: true, trim: true },
  contrasena: { type: String, required: true },
  rol: { type: String, enum: ['admin', 'cliente', 'online'], default: 'cliente' },
  haPagado: {
    type: Boolean,
    default: false
  },
  nombreGrupo: {
    type: String,
    trim: true,
    index: true, // Para que la ordenación sea más rápida
    default: null
  },

  esPremium: { type: Boolean, default: false },
  cancelaciones: { type: Number, default: 0 },
  tiposDeClases: { type: [String], enum: ['funcional', 'pilates', 'zumba'], required: true },
  avatar: { type: Object, default: {} },

  // --- Banderas de control del Admin ---
  incluyePlanDieta: { type: Boolean, default: false },
  incluyePlanEntrenamiento: { type: Boolean, default: false },

  // --- Inputs de Dieta ---
  dietaAlergias: { type: String, default: 'Ninguna' },
  dietaPreferencias: { type: String, default: 'Omnívoro, me gusta todo' },
  dietaComidas: { type: Number, default: 4 },
  historialMedico: { type: String, default: '' },
  horarios: { type: String, default: '' },
  platosFavoritos: { type: String, default: '' },
  dietaTiempoCocina: {
    type: String,
    enum: ['menos_15_min', '15_30_min', 'mas_30_min'],
    default: '15_30_min'
  },
  dietaHabilidadCocina: {
    type: String,
    enum: ['principiante', 'intermedio', 'avanzado'],
    default: 'intermedio'
  },
  dietaEquipamiento: {
    type: [String],
    default: ['basico']
  },
  dietaContextoComida: {
    type: String,
    enum: ['casa', 'oficina_tupper', 'restaurante'],
    default: 'casa'
  },
  dietaAlimentosOdiados: { type: String, default: 'Ninguno' },
  dietaRetoPrincipal: {
    type: String,
    enum: ['picoteo', 'social', 'organizacion', 'estres', 'raciones'],
    default: 'picoteo'
  },
  dietaBebidas: { type: String, default: 'Principalmente agua' },

  // --- ¡¡INPUTS DE ENTRENAMIENTO ACTUALIZADOS!! ---
  premiumMeta: {
    type: String,
    // (fuerza_pura = Powerlifting, rendimiento_atletico = Híbrido/Velocidad)
    enum: ['perder_grasa', 'hipertrofia', 'fuerza_pura', 'rendimiento_atletico', 'salud_general'],
    default: 'salud_general'
  },
  premiumFoco: { type: String, default: 'Cuerpo completo' },

  premiumEquipamiento: {
    type: [String], // Cambiado a Array de Strings
    default: ['solo_cuerpo']
    // Opciones: 'solo_cuerpo', 'bandas_elasticas', 'mancuernas_ligeras', 
    // 'mancuernas_ajustables', 'kettlebell', 'barra_dominadas', 'banco',
    // 'gym_basico', 'gym_completo'
  },
  premiumTiempo: { type: Number, default: 45 },

  premiumNivel: {
    type: String,
    enum: ['principiante_nuevo', 'intermedio_consistente', 'avanzado_programado'],
    default: 'principiante_nuevo'
  },
  premiumDiasSemana: {
    type: Number,
    default: 4
  },
  premiumLesiones: {
    type: String,
    default: 'Ninguna'
  },
  premiumEjerciciosOdiados: { // ¡NUEVO!
    type: String,
    default: 'Ninguno'
  },
  // -----------------------------------------

  // --- Datos Metabólicos (Dieta) ---
  genero: {
    type: String,
    enum: ['masculino', 'femenino'],
    default: 'masculino'
  },
  edad: {
    type: Number,
    min: 16,
    default: 25
  },
  altura: { // En centímetros
    type: Number,
    min: 100,
    default: 170
  },
  peso: { // En kilogramos
    type: Number,
    min: 40,
    default: 70
  },
  ocupacion: {
    type: String,
    enum: ['sedentaria', 'ligera', 'activa'],
    default: 'sedentaria'
  },
  ejercicio: {
    type: String,
    enum: ['0', '1-3', '4-5', '6-7'],
    default: '0'
  },
  objetivo: {
    type: String,
    enum: ['perder', 'mantener', 'ganar'],
    default: 'mantener'
  },
  kcalObjetivo: {
    type: Number,
    default: 2000
  },

}, { timestamps: true });

// Middleware para encriptar la contraseña antes de guardarla
esquemaUsuario.pre('save', async function (next) {
  if (!this.isModified('contrasena')) return next();
  const salt = await bcrypt.genSalt(10);
  this.contrasena = await bcrypt.hash(this.contrasena, salt);
  next();
});

// Middleware para borrar en cascada
esquemaUsuario.pre('deleteOne', { document: true, query: false }, async function (next) {
  try {
    const Clase = require('./Clase');

    // 1. Obtener todas las reservas del usuario para incrementar cupos
    const reservasDelUsuario = await Reserva.find({ usuario: this._id });

    // 2. Incrementar cuposDisponibles en cada clase donde tenía reserva
    for (const reserva of reservasDelUsuario) {
      await Clase.findByIdAndUpdate(
        reserva.clase,
        { $inc: { cuposDisponibles: 1 } }
      );
    }

    // 3. Eliminar datos relacionados
    await Promise.all([
      Reserva.deleteMany({ usuario: this._id }),
      PlanDieta.deleteMany({ usuario: this._id }),
      PlanEntrenamiento.deleteMany({ usuario: this._id }),
      Salud.deleteMany({ usuario: this._id }),
      // 4. Eliminar de todas las listas de espera
      Clase.updateMany(
        { listaEspera: this._id },
        { $pull: { listaEspera: this._id } }
      )
    ]);

    next();
  } catch (error) {
    next(error);
  }
});

// Método para verificar la contraseña
esquemaUsuario.methods.verificarContrasena = async function (contrasena) {
  return await bcrypt.compare(contrasena, this.contrasena);
};

const Usuario = mongoose.model('Usuario', esquemaUsuario);

module.exports = Usuario;

File: frontend/lib/models/usuario.dart
// models/usuario.dart
// ¡MODIFICADO! Actualizados los campos de entrenamiento (equipamiento, etc.)

import 'dart:convert';

class Usuario {
  final String id;
  final String nombre;
  final String correo;
  final String rol;
  final int cancelaciones;
  final List<String> tiposDeClases;
  final Map<String, dynamic> avatar; 
  
  // --- Datos Metabólicos (Dieta) ---
  final String genero;
  final int edad;
  final double altura;
  final double peso;
  final String ocupacion;
  final String ejercicio;
  final String objetivo;
  final int kcalObjetivo;

  // Campos Premium
  final bool esPremium;
  final DateTime? solicitudPremium;
  final bool incluyePlanDieta;
  final bool incluyePlanEntrenamiento;
  
  // --- Inputs Dieta (Base + Adherencia) ---
  final String dietaAlergias;
  final String dietaPreferencias;
  final int dietaComidas;
  final String historialMedico;
  final String horarios;
  final String platosFavoritos;
  final String dietaTiempoCocina;
  final String dietaHabilidadCocina;
  final List<String> dietaEquipamiento; // (Este ya estaba como Lista)
  final String dietaContextoComida;
  final String dietaAlimentosOdiados;
  final String dietaRetoPrincipal;
  final String dietaBebidas;
  
  // --- ¡INPUTS DE ENTRENAMIENTO ACTUALIZADOS! ---
  final String premiumMeta;
  final String premiumFoco;
  final List<String> premiumEquipamiento; // ¡CAMBIADO! (Era String)
  final int premiumTiempo;
  final String premiumNivel;
  final int premiumDiasSemana;
  final String premiumLesiones;
  final String premiumEjerciciosOdiados; // ¡NUEVO!
  // -------------------------------------------

  // Otros Campos
  final bool haPagado;
  final String? nombreGrupo; 

  Usuario({
    required this.id,
    required this.nombre,
    required this.correo,
    required this.rol,
    required this.cancelaciones,
    required this.tiposDeClases,
    required this.avatar,
    // Metabólicos
    required this.genero,
    required this.edad,
    required this.altura,
    required this.peso,
    required this.ocupacion,
    required this.ejercicio,
    required this.objetivo,
    required this.kcalObjetivo,
    // Premium
    required this.esPremium,
    required this.solicitudPremium,
    required this.incluyePlanDieta,
    required this.incluyePlanEntrenamiento,
    // Dieta
    required this.dietaAlergias,
    required this.dietaPreferencias,
    required this.dietaComidas,
    required this.historialMedico,
    required this.horarios,
    required this.platosFavoritos,
    required this.dietaTiempoCocina,
    required this.dietaHabilidadCocina,
    required this.dietaEquipamiento,
    required this.dietaContextoComida,
    required this.dietaAlimentosOdiados,
    required this.dietaRetoPrincipal,
    required this.dietaBebidas,
    
    // ¡ENTRENAMIENTO ACTUALIZADO!
    required this.premiumMeta,
    required this.premiumFoco,
    required this.premiumEquipamiento, // ¡CAMBIADO!
    required this.premiumTiempo,
    required this.premiumNivel,
    required this.premiumDiasSemana,
    required this.premiumLesiones,
    required this.premiumEjerciciosOdiados, // ¡NUEVO!

    required this.haPagado,
    this.nombreGrupo,
  });

  factory Usuario.fromJson(Map<String, dynamic> json) {
    // (Lógica de parseo de avatar sin cambios)
    Map<String, dynamic> parsedAvatar = {}; 
    if (json['avatar'] is Map<String, dynamic>) {
      parsedAvatar = json['avatar'];
    } else if (json['avatar'] is String) {
      try {
        var decoded = jsonDecode(json['avatar']);
        if (decoded is Map<String, dynamic>) {
          parsedAvatar = decoded;
        }
      } catch (e) {
        print('Warning: Could not parse avatar string: ${json['avatar']}');
      }
    }

    return Usuario(
      id: json['_id'] ?? '',
      nombre: json['nombre'] ?? '',
      correo: json['correo'] ?? '',
      rol: json['rol'] ?? 'cliente',
      cancelaciones: (json['cancelaciones'] as num?)?.toInt() ?? 0,
      tiposDeClases: (json['tiposDeClases'] as List<dynamic>?)?.cast<String>() ?? [],
      avatar: parsedAvatar, 
      
      // Metabólicos (Dieta)
      genero: json['genero'] ?? 'masculino',
      edad: (json['edad'] ?? 25).toInt(),
      altura: (json['altura'] ?? 170.0).toDouble(),
      peso: (json['peso'] ?? 70.0).toDouble(),
      ocupacion: json['ocupacion'] ?? 'sedentaria', 
      ejercicio: json['ejercicio'] ?? '0',           
      objetivo: json['objetivo'] ?? 'mantener',
      kcalObjetivo: (json['kcalObjetivo'] ?? 2000).toInt(),
      
      // Premium
      esPremium: json['esPremium'] ?? false,
      solicitudPremium: json['solicitudPremium'] != null 
          ? DateTime.tryParse(json['solicitudPremium'].toString()) 
          : null,
      incluyePlanDieta: json['incluyePlanDieta'] ?? false,
      incluyePlanEntrenamiento: json['incluyePlanEntrenamiento'] ?? false,
      
      // Dieta
      dietaAlergias: json['dietaAlergias'] ?? 'Ninguna',
      dietaPreferencias: json['dietaPreferencias'] ?? 'Omnívoro',
      dietaComidas: (json['dietaComidas'] ?? 4).toInt(),
      historialMedico: json['historialMedico'] ?? '',
      horarios: json['horarios'] ?? '',
      platosFavoritos: json['platosFavoritos'] ?? '',
      dietaTiempoCocina: json['dietaTiempoCocina'] ?? '15_30_min',
      dietaHabilidadCocina: json['dietaHabilidadCocina'] ?? 'intermedio',
      dietaEquipamiento: (json['dietaEquipamiento'] as List<dynamic>?)?.cast<String>() ?? ['basico'],
      dietaContextoComida: json['dietaContextoComida'] ?? 'casa',
      dietaAlimentosOdiados: json['dietaAlimentosOdiados'] ?? 'Ninguno',
      dietaRetoPrincipal: json['dietaRetoPrincipal'] ?? 'picoteo',
      dietaBebidas: json['dietaBebidas'] ?? 'Principalmente agua',

      // ¡ENTRENAMIENTO ACTUALIZADO!
      premiumMeta: json['premiumMeta'] ?? 'salud_general',
      premiumFoco: json['premiumFoco'] ?? 'Cuerpo completo',
      // ¡CAMBIADO! Lee una lista de strings
      premiumEquipamiento: (json['premiumEquipamiento'] as List<dynamic>?)?.cast<String>() ?? ['solo_cuerpo'],
      premiumTiempo: (json['premiumTiempo'] ?? 45).toInt(),
      premiumNivel: json['premiumNivel'] ?? 'principiante_nuevo',
      premiumDiasSemana: (json['premiumDiasSemana'] ?? 4).toInt(),
      premiumLesiones: json['premiumLesiones'] ?? 'Ninguna',
      premiumEjerciciosOdiados: json['premiumEjerciciosOdiados'] ?? 'Ninguno', // ¡NUEVO!
      
      // Otros
      haPagado: json['haPagado'] ?? false,
      nombreGrupo: json['nombreGrupo'], 
    );
  }
}